selfservice/strategy/code/strategy_verification.go

```diff
@@ -0,0 +1,409 @@
+func (s *Strategy) handleVerificationError(w http.ResponseWriter, r *http.Request, f *verification.Flow, body *submitSelfServiceVerificationFlowWithCodeMethodBody, err error) error {
+	if f != nil {
+		f.UI.SetCSRF(s.deps.GenerateCSRFToken(r))
+		f.UI.GetNodes().Upsert(
+			node.NewInputField("email", body.Email, node.CodeGroup, node.InputAttributeTypeEmail, node.WithRequiredInputAttribute).WithMetaLabel(text.NewInfoNodeInputEmail()),
+		)
+	}
+
+	return err
+}
+
+// swagger:model submitSelfServiceVerificationFlowWithLinkMethodBody
+type submitSelfServiceVerificationFlowWithCodeMethodBody struct {
+	// Email to Verify
+	//
+	// Needs to be set when initiating the flow. If the email is a registered
+	// verification email, a verification link will be sent. If the email is not known,
+	// a email with details on what happened will be sent instead.
+	//
+	// format: email
+	Email string `form:"email" json:"email"`
+
+	// Sending the anti-csrf token is only required for browser login flows.
+	CSRFToken string `form:"csrf_token" json:"csrf_token"`
+
+	// Method is the recovery method
+	//
+	// enum:
+	// - link
+	// - code
+	Method string `json:"method"`
+
+	// The id of the flow
+	Flow string `json:"flow" form:"flow"`
+
+	// The verification code
+	Code string `json:"code" form:"code"`
+}
+
+// getMethod returns the method of this submission or "" if no method could be found
+func (body *submitSelfServiceVerificationFlowWithCodeMethodBody) getMethod() string {
+	if body.Method != "" {
+		return body.Method
+	}
+	if body.Code != "" {
+		return verification.StrategyVerificationCodeName
+	}
+
+	return ""
+}
+
+func (s *Strategy) Verify(w http.ResponseWriter, r *http.Request, f *verification.Flow) (err error) {
+	body, err := s.decodeVerification(r)
+	if err != nil {
+		return s.handleVerificationError(w, r, nil, body, err)
+	}
+
+	if err := flow.MethodEnabledAndAllowed(r.Context(), s.VerificationStrategyID(), body.getMethod(), s.deps); err != nil {
+		return s.handleVerificationError(w, r, f, body, err)
+	}
+
+	if err := f.Valid(); err != nil {
+		return s.handleVerificationError(w, r, f, body, err)
+	}
+
+	switch f.State {
+	case verification.StateChooseMethod:
+		fallthrough
+	case verification.StateEmailSent:
+		return s.verificationHandleFormSubmission(w, r, f, body)
+	case verification.StatePassedChallenge:
+		return s.retryVerificationFlowWithMessage(w, r, f.Type, text.NewErrorValidationVerificationRetrySuccess())
+	default:
+		return s.retryVerificationFlowWithMessage(w, r, f.Type, text.NewErrorValidationVerificationStateFailure())
+	}
+}
+
+func (s *Strategy) createVerificationCodeForm(action string, code *string, email *string) *container.Container {
+	// re-initialize the UI with a "clean" new state
+	c := &container.Container{
+		Method: "POST",
+		Action: action,
+	}
+
+	c.Nodes.Append(
+		node.
+			NewInputField("code", code, node.CodeGroup, node.InputAttributeTypeNumber, node.WithRequiredInputAttribute).
```

Comment:
We wanted to make this a text type, also for recovery!

Reply:
Right, done.